modules/compiler/builtins/abacus/include/abacus/internal/horner_polynomial.h

```diff
@@ -20,14 +20,33 @@
 #include <abacus/abacus_config.h>
 #include <abacus/abacus_type_traits.h>
 
+namespace {
+template <typename T, typename E = typename TypeTraits<T>::ElementType>
+struct fmuladd_helper;
+
+#ifdef __CA_BUILTINS_HALF_SUPPORT
+template <typename T>
+struct fmuladd_helper<T, abacus_half> {
```

Comment:
[`fmuladd`](https://llvm.org/docs/LangRef.html#llvm-fmuladd-intrinsic) is the name of an LLVM intrinsic that specifically does not guarantee one way or the other whether it is implemented as `mul+add` or `fma`, but for half-precision, we need `fma`, I think it's confusing to use naming that indicates it might not be `fma`.

Reply:
I think we can avoid the helper entirely, actually. Now that we build in C++17 mode, we can use `if constexpr`.

Reply:
Good point, thanks. I was unsure specifically about whether for half-precision we needed an FMA or just needed a multiply+add with extra precision (though not necessarily exactly FMA-level precision).

Do you have a suggestion for a better name? Would "multiply_add" clear up confusion between that and `llvm.fmuladd`? Or would that still hide the fact we need FMA?

Reply:
> I think we can avoid the helper entirely, actually. Now that we build in C++17 mode, we can use `if constexpr`.

I tried that, but was seemingly not allowed to include `type_traits` (I was trying to use `std::is_same`)

Reply:
`if constexpr(sizeof(TypeTraits<T>::ElementType) == 2)` should work well enough for us. We could add more to `TypeTraits` to make this more readable too, if you prefer.

Reply:
Done